src/ui/CategoryFacet/CategoryFacetValues/CategoryFacetValues.ts

```diff
@@ -0,0 +1,142 @@
+import { CategoryFacet } from '../CategoryFacet';
+import { IFacetResponse, IFacetResponseValue } from '../../../rest/Facet/FacetResponse';
+import { CategoryFacetValue } from './CategoryFacetValue';
+import { FacetUtils } from '../../Facet/FacetUtils';
+import { $$ } from '../../../utils/Dom';
+import { find } from 'underscore';
+import { FacetValueState } from '../../../rest/Facet/FacetValueState';
+import { Utils } from '../../../utils/Utils';
+
+export class CategoryFacetValues {
+  private facetValues: CategoryFacetValue[];
+  private list = $$('ul', { className: 'coveo-dynamic-category-facet-values' }).el;
+
+  constructor(private facet: CategoryFacet) {
+    this.resetValues();
+  }
+
+  private formatDisplayValue(value: string) {
+    let returnValue = FacetUtils.tryToGetTranslatedCaption(<string>this.facet.options.field, value);
+    return this.facet.options.valueCaption[value] || returnValue;
+  }
+
+  public createFromResponse(response: IFacetResponse) {
+    this.facetValues = response.values.map(responseValue => this.createFacetValueFromResponse(responseValue));
+  }
+
+  private createFacetValueFromResponse(responseValue: IFacetResponseValue, path: string[] = []): CategoryFacetValue {
+    const newPath = [...path, responseValue.value];
+    const displayValue = FacetUtils.getDisplayValueFromValueCaption(
+      responseValue.value,
+      this.facet.options.field as string,
+      this.facet.options.valueCaption
+    );
+    const children = responseValue.children
+      ? responseValue.children.map(responseValue => this.createFacetValueFromResponse(responseValue, newPath))
+      : [];
+
+    return new CategoryFacetValue({
+      value: responseValue.value,
+      numberOfResults: responseValue.numberOfResults,
+      state: responseValue.state,
+      moreValuesAvailable: responseValue.moreValuesAvailable,
+      preventAutoSelect: false,
+      path: newPath,
+      displayValue,
+      children
+    }, this.facet);
+  }
+
+  public get allFacetValues() {
+    return this.facetValues;
+  }
+
+  public resetValues() {
+    this.facetValues = [];
+  }
+
+  private findValueWithPath(path: string[]) {
+    let facetValues = this.facetValues;
+    let facetValue: CategoryFacetValue;
+    let remainingPath = [...path];
+
+    do {
+      const value = remainingPath.shift();
+      facetValue = find(facetValues, facetValue => facetValue.value === value);
+
+      if (!facetValue) {
+        return null;
+      }
+      facetValues = facetValue.children;
+    } while (remainingPath.length);
+
+    return facetValue;
```

Comment:
This seems like a wasteful loop... 🤔 

Could `CategoryFacetValue` have a method `get fullValue()` which returns `path.join(delimiter) + value` and then do `find(facetValues, facetValue => facetValue.fullPath === path)` ?

Reply:
Facet values are in a tree like structure, not a single level list, the loops only "digs" into the branches level by level to check if the value is there.